File: packages/search/src/scoring.ts
import type { MatchQuality, MatchReason, MatchSource } from '@yanuka/types';

/**
 * Field weights, taken from the product spec.
 *
 * The ordering encodes an editorial judgement about identity: a name or a phone
 * number identifies a person, a city merely narrows them down. Notes score
 * lowest per hit because they are long and therefore easy to match by accident
 * — but they are the field that makes this database worth having, so they are
 * never excluded, only ranked below the precise signals.
 */
export const FIELD_WEIGHTS: Record<MatchSource, number> = {
  name: 100,
  phone: 100,
  alias: 90,
  email: 60,
  profession: 50,
  tag: 40,
  category: 40,
  organization: 35,
  specialty: 35,
  city: 30,
  country: 20,
  role: 30,
  notes: 20,
  reason_for_saving: 25,
};

/**
 * Multiplier for how well the term matched.
 *
 * `exact` is the whole field equalling the term; `prefix` is a completion while
 * the user is still typing; `fulltext` is a token hit somewhere inside a longer
 * field; `fuzzy` is an edit-distance rescue and is deliberately expensive.
 */
export const QUALITY_MULTIPLIERS: Record<MatchQuality, number> = {
  exact: 1,
  prefix: 0.8,
  fulltext: 0.55,
  fuzzy: 0.45,
};

/** Small additive nudges applied once per contact, not per match. */
export const BONUSES = {
  /** The record is complete enough to act on. */
  hasPhone: 3,
  hasOrganization: 2,
  favorite: 12,
  /** Recently opened records are usually the ones being worked on. */
  recentlyViewed: 8,
} as const;

/** Score for a single match, before per-contact bonuses. */
export function scoreMatch(
  source: MatchSource,
  quality: MatchQuality,
  /** Extra discount, e.g. PROCLITIC_PENALTY for a grammatically-guessed stem. */
  penalty = 1,
): number {
  return FIELD_WEIGHTS[source] * QUALITY_MULTIPLIERS[quality] * penalty;
}

export interface ScoreInputs {
  reasons: MatchReason[];
  hasPhone: boolean;
  hasOrganization: boolean;
  isFavorite: boolean;
  /** Milliseconds since the contact was last opened, or null if never. */
  msSinceViewed: number | null;
}

const RECENT_VIEW_WINDOW_MS = 7 * 24 * 60 * 60 * 1000;

/**
 * Combine per-match scores into a single ranking score.
 *
 * Matches on the *same* field do not simply add up: five hits of the same query
 * word inside one long note should not outrank a single exact name match. Each
 * field therefore contributes its best match at full value and any additional
 * matches at a steep discount, which rewards a contact matching across several
 * different fields — the real signal that a result is what the user meant.
 */
export function combineScore(inputs: ScoreInputs): number {
  const bestPerField = new Map<MatchSource, number[]>();

  for (const reason of inputs.reasons) {
    const bucket = bestPerField.get(reason.source);
    if (bucket) bucket.push(reason.score);
    else bestPerField.set(reason.source, [reason.score]);
  }

  let total = 0;
  for (const scores of bestPerField.values()) {
    scores.sort((a, b) => b - a);
    total += scores[0]!;
    for (let i = 1; i < scores.length; i += 1) {
      total += scores[i]! * 0.15;
    }
  }

  if (inputs.hasPhone) total += BONUSES.hasPhone;
  if (inputs.hasOrganization) total += BONUSES.hasOrganization;
  if (inputs.isFavorite) total += BONUSES.favorite;
  if (inputs.msSinceViewed != null && inputs.msSinceViewed < RECENT_VIEW_WINDOW_MS) {
    // Linear decay across the window rather than a cliff at day seven.
    total += BONUSES.recentlyViewed * (1 - inputs.msSinceViewed / RECENT_VIEW_WINDOW_MS);
  }

  return Math.round(total * 100) / 100;
}

/**
 * Map an FTS5 bm25 rank onto a `fulltext` quality multiplier.
 *
 * bm25 values are only comparable within a single query, so they are used to
 * order *within* the full-text layer and never compared against the exact or
 * fuzzy layers. `rank` is bm25's output (more negative is better) and
 * `worstRank` the least relevant value in the same result set.
 */
export function bm25ToQualityFactor(rank: number, bestRank: number, worstRank: number): number {
  if (!Number.isFinite(rank)) return 0.5;
  const span = worstRank - bestRank;
  if (span <= 0) return 1;
  const position = (worstRank - rank) / span; // 1 at the best rank, 0 at the worst
  return 0.6 + 0.4 * position;
}

/** Minimum similarity for a fuzzy candidate to be shown at all. */
export const FUZZY_MIN_SIMILARITY = 0.72;

/** Maximum edit distance considered, regardless of term length. */
export const FUZZY_MAX_DISTANCE = 2;

/**
 * Quality factor for a fuzzy hit, scaled by how close the match actually was.
 * A distance-1 typo ranks well above a distance-2 guess.
 */
export function fuzzyQualityFactor(similarityScore: number): number {
  if (similarityScore >= 1) return 1;
  if (similarityScore < FUZZY_MIN_SIMILARITY) return 0;
  const range = 1 - FUZZY_MIN_SIMILARITY;
  return 0.35 + 0.35 * ((similarityScore - FUZZY_MIN_SIMILARITY) / range);
}

File: crates/yanuka-search/src/scoring.rs
//! Ranking weights, mirrored from `packages/search/src/scoring.ts`.
//!
//! The conformance tests assert result *ordering* rather than absolute scores:
//! two implementations agreeing on the order is what users experience, and
//! comparing floats across languages is brittle for no benefit.

use serde::{Deserialize, Serialize};

/// Which field produced a match.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum MatchSource {
    Name,
    Phone,
    Alias,
    Email,
    Profession,
    Role,
    Tag,
    Category,
    Organization,
    Specialty,
    City,
    Country,
    Notes,
    ReasonForSaving,
}

/// How well the term matched, which sets the multiplier.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum MatchQuality {
    Exact,
    Prefix,
    Fulltext,
    Fuzzy,
}

/// Field weights.
///
/// A name or a phone number identifies a person; a city merely narrows them
/// down. Notes score lowest per hit because they are long and therefore easy to
/// match by accident — but they are never excluded, because they are what makes
/// this archive worth keeping.
pub fn field_weight(source: MatchSource) -> f64 {
    match source {
        MatchSource::Name | MatchSource::Phone => 100.0,
        MatchSource::Alias => 90.0,
        MatchSource::Email => 60.0,
        MatchSource::Profession => 50.0,
        MatchSource::Tag | MatchSource::Category => 40.0,
        MatchSource::Organization | MatchSource::Specialty => 35.0,
        MatchSource::City | MatchSource::Role => 30.0,
        MatchSource::ReasonForSaving => 25.0,
        MatchSource::Country => 20.0,
        MatchSource::Notes => 20.0,
    }
}

pub fn quality_multiplier(quality: MatchQuality) -> f64 {
    match quality {
        MatchQuality::Exact => 1.0,
        MatchQuality::Prefix => 0.8,
        MatchQuality::Fulltext => 0.55,
        MatchQuality::Fuzzy => 0.45,
    }
}

/// Score for one match, before per-contact bonuses.
pub fn score_match(source: MatchSource, quality: MatchQuality, penalty: f64) -> f64 {
    field_weight(source) * quality_multiplier(quality) * penalty
}

pub const BONUS_HAS_PHONE: f64 = 3.0;
pub const BONUS_HAS_ORGANIZATION: f64 = 2.0;
pub const BONUS_FAVORITE: f64 = 12.0;
pub const BONUS_RECENTLY_VIEWED: f64 = 8.0;

/// Minimum similarity for a fuzzy candidate to be shown at all.
pub const FUZZY_MIN_SIMILARITY: f64 = 0.72;

/// Maximum edit distance considered, regardless of term length.
pub const FUZZY_MAX_DISTANCE: usize = 2;

/// Map an FTS5 bm25 rank onto a `Fulltext` quality factor.
///
/// bm25 values are comparable only within one query, so they order results
/// *within* the full-text layer and are never compared against the exact or
/// fuzzy layers.
pub fn bm25_to_quality_factor(rank: f64, best_rank: f64, worst_rank: f64) -> f64 {
    if !rank.is_finite() {
        return 0.5;
    }
    let span = worst_rank - best_rank;
    if span <= 0.0 {
        return 1.0;
    }
    let position = (worst_rank - rank) / span;
    0.6 + 0.4 * position
}

/// Quality factor for a fuzzy hit, scaled by how close the match was.
pub fn fuzzy_quality_factor(similarity: f64) -> f64 {
    if similarity >= 1.0 {
        return 1.0;
    }
    if similarity < FUZZY_MIN_SIMILARITY {
        return 0.0;
    }
    let range = 1.0 - FUZZY_MIN_SIMILARITY;
    0.35 + 0.35 * ((similarity - FUZZY_MIN_SIMILARITY) / range)
}

/// Damerau-Levenshtein distance with an early-exit ceiling.
///
/// Only ever run over candidates the trigram index already narrowed down, never
/// over the whole table.
pub fn edit_distance(a: &str, b: &str, max_distance: usize) -> usize {
    let a: Vec<char> = a.chars().collect();
    let b: Vec<char> = b.chars().collect();

    if a == b {
        return 0;
    }
    if a.len().abs_diff(b.len()) > max_distance {
        return max_distance + 1;
    }
    if a.is_empty() {
        return b.len();
    }
    if b.is_empty() {
        return a.len();
    }

    let mut prev_prev: Vec<usize> = vec![0; b.len() + 1];
    let mut prev: Vec<usize> = (0..=b.len()).collect();
    let mut current: Vec<usize> = vec![0; b.len() + 1];

    for i in 1..=a.len() {
        current[0] = i;
        let mut row_min = current[0];

        for j in 1..=b.len() {
            let cost = usize::from(a[i - 1] != b[j - 1]);
            let mut value = (current[j - 1] + 1).min(prev[j] + 1).min(prev[j - 1] + cost);

            // Transposition: `ab` -> `ba` costs one, not two. This is what makes
            // it Damerau rather than plain Levenshtein, and it matters for typos.
            if i > 1 && j > 1 && a[i - 1] == b[j - 2] && a[i - 2] == b[j - 1] {
                value = value.min(prev_prev[j - 2] + cost);
            }

            current[j] = value;
            row_min = row_min.min(value);
        }

        if row_min > max_distance {
            return max_distance + 1;
        }

        std::mem::swap(&mut prev_prev, &mut prev);
        std::mem::swap(&mut prev, &mut current);
    }

    prev[b.len()]
}

/// Similarity in [0, 1], normalized by the longer string so a one-character
/// typo costs more in a short word than in a long one.
pub fn similarity(a: &str, b: &str) -> f64 {
    let longest = a.chars().count().max(b.chars().count());
    if longest == 0 {
        return 1.0;
    }
    let distance = edit_distance(a, b, longest);
    1.0 - (distance as f64 / longest as f64)
}

/// Character trigrams, padded so short strings still produce grams.
pub fn trigrams(value: &str) -> Vec<String> {
    if value.is_empty() {
        return Vec::new();
    }
    let padded: Vec<char> = format!(" {value} ").chars().collect();
    padded.windows(3).map(|w| w.iter().collect()).collect()
}

/// How many trigrams a candidate must share before it is worth scoring.
/// Forty percent overlap tolerates roughly one edit per five characters.
pub fn min_shared_trigrams(term_trigram_count: usize) -> usize {
    ((term_trigram_count as f64 * 0.4).ceil() as usize).max(1)
}
